crates/matrix-sdk-crypto/src/backups/mod.rs

```diff
@@ -105,46 +168,95 @@ impl BackupMachine {
+    /// Check if our own device has signed the given JSON `Value`.
+    async fn check_own_device_signature(&self, auth_data: Value) -> SignatureState {
+        match self.account.is_signed(auth_data) {
+            Ok(_) => SignatureState::Trusted,
+            Err(e) => match e {
+                crate::SignatureError::NoSignatureFound => SignatureState::Missing,
+                _ => SignatureState::Invalid,
+            },
         }
+    }
+
+    /// Check if our own cross-signing user identity has signed the given JSON
+    /// `Value`.
+    async fn check_own_identity_signature(
+        &self,
+        auth_data: Value,
+    ) -> Result<SignatureState, CryptoStoreError> {
+        let user_id = self.account.user_id();
+        let identity = self.store.get_identity(user_id).await?;
+
+        let ret = if let Some(identity) = identity.and_then(|i| i.own()) {
+            match identity.master_key().is_signed_by(auth_data) {
+                Ok(_) => {
+                    if identity.is_verified() {
+                        SignatureState::Trusted
+                    } else {
+                        SignatureState::Valid
+                    }
+                }
+                Err(e) => match e {
+                    crate::SignatureError::NoSignatureFound => SignatureState::Missing,
+                    _ => SignatureState::Invalid,
+                },
+            }
+        } else {
+            SignatureState::Missing
+        };
 
-        let auth_data: AuthData = serde_json::from_value(serialized_auth_data.clone())?;
+        Ok(ret)
+    }
 
-        trace!(?auth_data, "Verifying backup auth data");
+    /// Check if the given JSON `Value` has been signed by the given `Device`
+    /// with the key that belongs to the given `DeviceKeyId`.
+    fn backup_signed_by_device(
+        &self,
+        device_key_id: &DeviceKeyId,
+        device: Option<Device>,
+        auth_data: Value,
+    ) -> SignatureState {
+        if let Some(device) = device {
+            if device.is_signed_by_device(auth_data).is_ok() {
+                if device.verified() {
+                    SignatureState::Trusted
+                } else {
+                    SignatureState::Valid
+                }
+            } else {
+                SignatureState::Invalid
+            }
+        } else {
+            trace!(
+                device_id = device_key_id.device_id().as_str(),
+                "Device not found, can't check signature"
+            );
+            SignatureState::Missing
+        }
+    }
+
+    /// Check if the JSON `Value` has been signed by any of our devices.
+    ///
+    /// This method will abort as soon as it finds a signature from a trusted
+    /// device if the `check_all` argument is set to `false`. If the `check_all`
+    /// argument is set to `true`, the method will check and find the
+    /// `SignatureState` for all the signatures that were given.
+    async fn backup_signed_by_any_trusted_device(
```

Comment:
This is likely misnamed since we're not actually answering the question "Is this backup signed by *any* trusted device?". Instead, we're computing the signature state for either all signatures or until we've encountered the first signature made by a trusted device.

I think the method should be named `test_device_signatures` and the boolean parameter something like `stop_on_first_trusted`.

Reply:
Fixed by https://github.com/matrix-org/matrix-rust-sdk/pull/725/commits/132bf7bd7aebd320ee5c3658a5296c238a26e7c2.